packages/firestore/test/integration/api/transactions.test.ts

```diff
@@ -23,7 +23,184 @@ import * as integrationHelpers from '../util/helpers';
 
 const apiDescribe = integrationHelpers.apiDescribe;
 
-apiDescribe('Database transactions', (persistence: boolean) => {
+apiDescribe.only('Database transactions', (persistence: boolean) => {
+  type TransactionStep = (
+    transaction: firestore.Transaction,
+    docRef: firestore.DocumentReference
+  ) => void;
+
+  /**
+   * Used for testing that all possible combinations of executing transactions
+   * result in the desired document value or error.
+   *
+   * The transaction steps are postfixed by numbers to indicate the calling
+   * order. For example, calling `set1()` followed by `set2()` should result in
+   * the document being set to the value specified by `set2()`.
+   */
+  class TransactionTester {
+    constructor(readonly db: firestore.FirebaseFirestore) {}
+
+    async delete(
+      transaction: firestore.Transaction,
+      docRef: firestore.DocumentReference
+    ): Promise<void> {
+      transaction.delete(docRef);
+    }
+
+    async update1(
+      transaction: firestore.Transaction,
+      docRef: firestore.DocumentReference
+    ): Promise<void> {
+      transaction.update(docRef, { foo: 'bar1' });
+    }
+
+    async update2(
+      transaction: firestore.Transaction,
+      docRef: firestore.DocumentReference
+    ): Promise<void> {
+      transaction.update(docRef, { foo: 'bar2' });
+    }
+
+    async set1(
+      transaction: firestore.Transaction,
+      docRef: firestore.DocumentReference
+    ): Promise<void> {
+      transaction.set(docRef, { foo: 'bar1' });
+    }
+
+    async set2(
+      transaction: firestore.Transaction,
+      docRef: firestore.DocumentReference
+    ): Promise<void> {
+      transaction.set(docRef, { foo: 'bar2' });
+    }
+
+    async getExisting(
+      transaction: firestore.Transaction,
+      docRef: firestore.DocumentReference
+    ): Promise<void> {
+      return transaction.get(docRef).then(docSnap => {
+        expect(docSnap.exists).to.equal(true);
+      });
+    }
+
+    async getMissing(
+      transaction: firestore.Transaction,
+      docRef: firestore.DocumentReference
+    ): Promise<void> {
+      return transaction.get(docRef).then(docSnap => {
+        expect(docSnap.exists).to.equal(false);
+      });
+    }
+
+    /**
+     * Validates that executing the provided steps in a transaction results in
+     * the doc value equal to the expected value.
+     *
+     * @param expected The expected value of the doc.
+     * @param docExists Whether the document the transaction is being run on
+     * should exist beforehand.
+     * @param stages The series of transaction steps to execute inside the
+     * transaction.
+     */
+    async validateSequenceSuccess(
+      expected: object | undefined,
+      docExists: boolean,
+      ...stages: TransactionStep[]
```

Comment:
You're calling these `TransactionStep` but the variable here is `stages`. Consider naming these consistently. Otherwise it prompts the question of how is a stage different from a step?